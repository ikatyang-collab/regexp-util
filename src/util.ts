import { Base } from './base.js'

export const wrap = (input: string | Base) =>
  typeof input !== 'string' || input.length === 1
    ? input.toString()
    : `(?:${input})`

export const concat = (...inputs: Array<string | Base>) => inputs.join('')

export const optional = (input: string | Base) => `${wrap(input)}?`

export const repeat = (input: string | Base, count: number, max?: number) =>
  max === undefined
    ? `${wrap(input)}{${count}}`
    : max === Infinity
    ? count > 1
      ? `${wrap(input)}{${count},}`
      : count < 1
      ? `${wrap(input)}*`
      : `${wrap(input)}+`
    : `${wrap(input)}{${count},${max}}`
