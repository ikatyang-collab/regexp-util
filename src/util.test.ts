import { test, expect } from 'vitest'
import { concat, optional, repeat, wrap } from './util.js'

test('wrap: accept char', () => {
  expect(wrap('a')).toMatchSnapshot()
})

test('wrap: accept string', () => {
  expect(wrap('abc')).toMatchSnapshot()
})

test('concat', () => {
  expect(concat('hello', 'world')).toMatchSnapshot()
})

test('optional', () => {
  expect(optional('abc')).toMatchSnapshot()
})

test('repeat: count', () => {
  expect(repeat('abc', 1)).toMatchSnapshot()
})

test('repeat: count, max != Infinity', () => {
  expect(repeat('abc', 1, 2)).toMatchSnapshot()
})

test('repeat: count = 0, max = Infinity', () => {
  expect(repeat('abc', 0, Infinity)).toMatchSnapshot()
})

test('repeat: count = 1, max = Infinity', () => {
  expect(repeat('abc', 1, Infinity)).toMatchSnapshot()
})

test('repeat: count > 1, max = Infinity', () => {
  expect(repeat('abc', 2, Infinity)).toMatchSnapshot()
})
