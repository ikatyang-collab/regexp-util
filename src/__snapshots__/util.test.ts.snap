// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`concat 1`] = `"helloworld"`;

exports[`optional 1`] = `"(?:abc)?"`;

exports[`repeat: count = 0, max = Infinity 1`] = `"(?:abc)*"`;

exports[`repeat: count = 1, max = Infinity 1`] = `"(?:abc)+"`;

exports[`repeat: count > 1, max = Infinity 1`] = `"(?:abc){2,}"`;

exports[`repeat: count 1`] = `"(?:abc){1}"`;

exports[`repeat: count, max != Infinity 1`] = `"(?:abc){1,2}"`;

exports[`wrap: accept char 1`] = `"a"`;

exports[`wrap: accept string 1`] = `"(?:abc)"`;
